fleetspeak/src/server/https/client_certificate.go

```diff
@@ -1,32 +1,87 @@
-func GetClientCert(req *http.Request, hn string, frontendMode cpb.FrontendMode) (*x509.Certificate, error) {
+func GetClientCert(req *http.Request, hn string, frontendMode cpb.FrontendMode, chcksumHeader string) (*x509.Certificate, error) {
 	switch frontendMode {
 	case cpb.FrontendMode_MTLS:
-		if hn == "" {
+		if hn == "" && chcksumHeader == "" {
 			return getCertFromTLS(req)
 		}
 	case cpb.FrontendMode_HEADER_TLS:
-		if hn != "" {
-			return getCertFromHeader(hn, req.Header)
+		if hn != "" && chcksumHeader == "" {
+			return getCertFromHeader(hn, req.Header, "", false)
+		}
+	case cpb.FrontendMode_HEADER_TLS_CHECKSUM:
+		if (hn != "" && chcksumHeader != "") {
+			return getCertFromHeader(hn, req.Header, chcksumHeader, true)
 		}
 	}
-	return nil, fmt.Errorf("received invalid frontend mode combination: frontendMode=%s, clientCertHeader=%s", frontendMode, hn)
+	return nil, fmt.Errorf("received invalid frontend mode combination: frontendMode=%s, clientCertHeader=%s, clientCertHeaderChecksum=%s",
+				frontendMode, hn, chcksumHeader)
+}
+
+func calcClientCertSha256(clientCert string) (string) {
+	// Decode the PEM string
+	block, _ := pem.Decode([]byte(clientCert))
+	if block == nil {
+		log.Warningln("Failed to decode PEM certificate")
+		return ""
+	}
+	// Calculate the SHA-256 digest of the DER certificate
+	sha256Digest := sha256.Sum256(block.Bytes)
+
+	// Convert the SHA-256 digest to a hexadecimal string
+	sha256HexStr := fmt.Sprintf("%x", sha256Digest)
+
+	sha256Binary, err := hex.DecodeString(sha256HexStr)
+	if err != nil {
+		log.Warningf("error decoding hexdump: %v\n", err)
+		return ""
+	}
+
+	// Convert the hexadecimal string to a base64 encoded string
+	// It also removes trailing "=" padding characters
+	base64EncodedStr := strings.TrimRight(base64.StdEncoding.EncodeToString(sha256Binary), "=")
+
+	// Rreturn the base64 encoded string
+	return base64EncodedStr
 }
 
-func getCertFromHeader(hn string, rh http.Header) (*x509.Certificate, error) {
+func verifyCertSha256Fingerprint(chcksumHeader string, rh http.Header, headerCert string) (error) {
+	//fmt.Println("--------------------------- received cert in header")
+	clientCertSha256Fingerprint := rh.Get(chcksumHeader)
+	if clientCertSha256Fingerprint != "" {
+		//fmt.Println("----- received client_cert_sha256_fingerprint:")
+		//fmt.Println(clientCertSha256Fingerprint)
+		calculatedClientCertSha256 := calcClientCertSha256(headerCert)
+		//fmt.Println("----- calculated client cert sha256 fingerprint:")
+		//fmt.Println(calculatedClientCertSha256)
+		if (calculatedClientCertSha256 != clientCertSha256Fingerprint) {
+			return errors.New("received client certificate checksum is invalid")
+		}
+	} else {
+		return errors.New("no client certificate checksum received in header")
+	}
+	return nil
+}
+
+func getCertFromHeader(hn string, rh http.Header, chcksumHeader string, verifyFingerprint bool) (*x509.Certificate, error) {
```

Comment:
I'd leave the verification out of the scope of this function as it's not relevant to all use cases of getCertFromHeader. I'd make users of this function (like GetClientCert above) do the verification themselves.

Reply:
Moved the verification code outside the scope of this function.